File: packages/guidance/src/policies/noForOfLoops.ts
import { makeFindings } from "@better-typescript/core/engine/policy"
import type { Guidance } from "@better-typescript/core/engine/policy/data"
import {
  noForOfLoopsMatcher,
  type NoForOfLoopsFact
} from "@better-typescript/matchers/builtins/noForOfLoops"
import { makeBuiltinPolicy } from "../definePolicy.js"

const synchronousHint =
  "Use Effect's Array module, such as Array.map(), Array.reduce(), " +
  "Array.filter(), or Array.flatMap(), instead."

const asynchronousHint =
  "Use Stream.fromAsyncIterable(...).pipe(Stream.map(...), Stream.runCollect) or another " +
  "Stream/Effect combinator instead; Array combinators do not consume AsyncIterable values."

const noForOfLoopsGuidance: Guidance<NoForOfLoopsFact> = () => (match) =>
  makeFindings(
    match.target,
    "Avoid imperative logic in for..of loops.",
    match.fact.isAsync ? asynchronousHint : synchronousHint,
    match.fact
  )

export const noForOfLoops = makeBuiltinPolicy(
  "no-for-of-loops",
  noForOfLoopsMatcher,
  noForOfLoopsGuidance
)

File: internal/rules/no_for_of_loops/no_for_of_loops.go
package no_for_of_loops

import (
	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/microsoft/typescript-go/shim/ast"
)

var synchronousMessage = rule.RuleMessage{
	Id:          "noForOfLoops",
	Description: "Avoid imperative logic in for..of loops.",
	Help:        "Use Effect's Array module, such as Array.map(), Array.reduce(), Array.filter(), or Array.flatMap(), instead.",
}

var asynchronousMessage = rule.RuleMessage{
	Id:          "noForOfLoops",
	Description: "Avoid imperative logic in for..of loops.",
	Help:        "Use Stream.fromAsyncIterable(...).pipe(Stream.map(...), Stream.runCollect) or another Stream/Effect combinator instead; Array combinators do not consume AsyncIterable values.",
}

var NoForOfLoopsRule = rule.Rule{
	Name: "no-for-of-loops",
	Run: func(ctx rule.RuleContext, options any) rule.RuleListeners {
		return rule.RuleListeners{ast.KindForOfStatement: func(node *ast.Node) {
			message := synchronousMessage
			if node.AsForInOrOfStatement().AwaitModifier != nil {
				message = asynchronousMessage
			}
			ctx.ReportNode(node, message)
		}}
	},
}
